searx/network/context.py

```diff
@@ -0,0 +1,311 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+# lint: pylint
+# pyright: basic
+# pylint: disable=missing-class-docstring
+"""Deal with 
+
+* total timeout
+* record HTTP runtime
+* retry strategy
+
+```python
+context = NetworkContextRetryFunction(...)  # or another implementation
+
+def my_engine():
+    http_client = context.get_http_client()
+    ip_ifconfig = http_client.request("GET", "https://ifconfig.me/")
+    print("ip from ifconfig.me ", ip_ifconfig)
+    ip_myip = http_client.request("GET", "https://api.myip.com").json()["ip"]
+    print("ip from api.myip.com", ip_myip)
+    assert ip_ifconfig == ip_myip  # always true with NetworkContextRetryFunction and NetworkContextRetrySameHTTPClient
+
+result = context.call(my_engine)
+print('HTTP runtime:', context.get_total_time())
+```
+
+Various implementations define what to do when there is an exception in the function `my_engine`:
+* `NetworkContextRetryFunction` gets another HTTP client, and try again the whole function.
+* `NetworkContextRetryDifferentHTTPClient` gets another HTTP client, and try again the query.
+* `NetworkContextRetrySameHTTPClient` try again the query with the same HTTP client.
+"""
+import functools
+import ssl
+import threading
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from timeit import default_timer
+from typing import Callable, Optional, final
+
+try:
+    from typing import ParamSpec, TypeVar
+except ImportError:
+    # to support Python < 3.10
+    from typing_extensions import ParamSpec, TypeVar
+
+import httpx
+
+from searx.network.client import (
+    ABCHTTPClient,
+    SoftRetryHTTPException,
+)
+
+P = ParamSpec('P')
+R = TypeVar('R')
+HTTPCLIENTFACTORY = Callable[[], ABCHTTPClient]
+
+
+THREADLOCAL = threading.local()
```

Comment:
Since NetworkContext is per user query and per engine, THREADLOCAL can be removed (?)

Reply:
Done